src/geometry/manifold/derivation_bundle.lean

```diff
@@ -0,0 +1,153 @@
+/-
+Copyright © 2020 Nicolò Cavalleri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Nicolò Cavalleri
+-/
+
+import geometry.manifold.algebra.smooth_functions
+import ring_theory.derivation
+
+/-!
+
+# Derivation bundle
+
+In this file we define the derivations at a point of a manifold on the algebra of smooth fuctions.
+Moreover we define the total bundle of derivations (although at the moment it has not been given a
+topology). Finally we define the differential of a function in terms of derivations.
+
+The content of this file is not meant to be regarded as an alternative definition to the current
+tangent bundle but rather as a purely algebraic theory that provides a purely algebraic definition
+of the Lie algebra for a Lie group.
+
+-/
+
+variables (𝕜 : Type*) [nondiscrete_normed_field 𝕜]
+{E : Type*} [normed_group E] [normed_space 𝕜 E]
+{H : Type*} [topological_space H] (I : model_with_corners 𝕜 E H)
+(M : Type*) [topological_space M] [charted_space H M] (n : with_top ℕ)
+
+open_locale manifold
+
+-- the following two instances prevent poorly understood type class inference timeout problems
+instance smooth_functions_algebra : algebra 𝕜 C^∞⟮I, M; 𝕜⟯ := by apply_instance
+instance smooth_functions_tower : is_scalar_tower 𝕜 C^∞⟮I, M; 𝕜⟯ C^∞⟮I, M; 𝕜⟯ := by apply_instance
+
+/-- Type synonym, introduced to put a different `has_scalar` action on `C^n⟮I, M; 𝕜⟯`
+which is defined as `f • r = f(x) * r`. -/
+@[nolint unused_arguments] def pointed_smooth_map (x : M) := C^n⟮I, M; 𝕜⟯
+
+localized "notation `C^` n `⟮` I `,` M `;` 𝕜 `⟯⟨` x `⟩` :=
+  pointed_smooth_map 𝕜 I M n x" in derivation
+
+variables {𝕜 M}
+
+namespace pointed_smooth_map
+
+instance {x : M} : has_coe_to_fun C^∞⟮I, M; 𝕜⟯⟨x⟩ := by { rw pointed_smooth_map, apply_instance }
+instance {x : M} : comm_ring C^∞⟮I, M; 𝕜⟯⟨x⟩ := by { rw pointed_smooth_map, apply_instance }
+instance {x : M} : algebra 𝕜 C^∞⟮I, M; 𝕜⟯⟨x⟩ := by { dunfold pointed_smooth_map, apply_instance }
+instance {x : M} : inhabited C^∞⟮I, M; 𝕜⟯⟨x⟩ := ⟨0⟩
+
+instance {x : M} : algebra C^∞⟮I, M; 𝕜⟯⟨x⟩ C^∞⟮I, M; 𝕜⟯ :=
+by { dunfold pointed_smooth_map, apply_instance }
+
+instance {x : M} : is_scalar_tower 𝕜 C^∞⟮I, M; 𝕜⟯⟨x⟩ C^∞⟮I, M; 𝕜⟯ :=
+by { dunfold pointed_smooth_map, apply_instance }
```

Comment:
I think better to avoid `dunfold` here as it introduces `id`s that might confuse typeclass search
```suggestion
instance {x : M} : algebra C^∞⟮I, M; 𝕜⟯⟨x⟩ C^∞⟮I, M; 𝕜⟯ := algebra.id C^∞⟮I, M; 𝕜⟯
instance {x : M} : is_scalar_tower 𝕜 C^∞⟮I, M; 𝕜⟯⟨x⟩ C^∞⟮I, M; 𝕜⟯ := is_scalar_tower.right
```